crates/sui-indexer-alt/src/handlers/mod.rs

```diff
@@ -0,0 +1,425 @@
+                        elapsed_ms = elapsed * 1000.0,
+                        "Processed checkpoint",
+                    );
+
+                    metrics
+                        .total_handler_checkpoints_processed
+                        .with_label_values(&[H::NAME])
+                        .inc();
+
+                    metrics
+                        .total_handler_rows_created
+                        .with_label_values(&[H::NAME])
+                        .inc_by(values.len() as u64);
+
+                    tx.send(Indexed {
+                        sequence_number: checkpoint.checkpoint_summary.sequence_number,
+                        values,
+                    })
+                    .await
+                    .map_err(|_| Break::Cancel)?;
+
+                    Ok(())
+                }
+            })
+            .await
+        {
+            Ok(()) | Err(Break::Cancel) => {
+                info!(pipeline = H::NAME, "Shutdown received, stopping handler");
+            }
+
+            Err(Break::Err(e)) => {
+                error!(pipeline = H::NAME, "Error from handler: {e}");
+                cancel.cancel();
+            }
+        };
+    })
+}
+
+/// The commiter task is responsible for gathering rows to write to the database. It is a single
+/// work loop that gathers checkpoint-wise row information, and periodically writes them out to the
+/// database.
+///
+/// The period between writes is controlled by the following factors:
+///
+/// - Time since the last write (controlled by `config.commit_interval`). If there are rows pending
+///   and this interval has elapsed since the last attempted commit, the committer will attempt
+///   another write.
+///
+/// - Time since last attempted write (controlled by `COOLDOWN_INTERVAL` and `RETRY_INTERVAL`). If
+///   there was a recent successful write, the next write will wait at least `COOLDOWN_INTERVAL`,
+///   and if there was a recent unsuccessful write, the next write will wait at least
+///   `RETRY_INTERVAL`. This is to prevent one committer from hogging the database.
+///
+/// - Number of pending rows. If this exceeds `H::BATCH_SIZE`, the committer will attempt to write
+///   out at most `H::CHUNK_SIZE` worth of rows to the DB.
+///
+/// If a write fails, the commiter will save the batch it meant to write and try to write them
+/// again at the next opportunity, potentially adding more rows to the batch if more have arrived
+/// in the interim.
+///
+/// This task will shutdown if canceled via the `cancel` token, or if the channel it receives data
+/// on has been closed by the handler for some reason.
+fn committer<H: Handler + 'static>(
+    db: Db,
+    mut rx: mpsc::Receiver<Indexed<H>>,
+    config: CommitterConfig,
+    metrics: Arc<IndexerMetrics>,
+    cancel: CancellationToken,
+) -> JoinHandle<()> {
+    spawn_monitored_task!(async move {
+        // The `poll` interval controls the maximum time to wait between commits, regardless of the
```

Comment:
Curious why do we need this. Wouldn't we want to commit the batch as soon as enough are available from handler?

Reply:
I think I ended up covering this while responding to the previous comment -- we do commit a chunk as soon as enough is available, as long as we haven't attempted to push out a chunk in the very near past (last 20ms).